### PR TITLE
CB-17015 Ephemeral Disk in Data Mart and Real-time Data Mart Data Hub Clusters…

### PR DESCRIPTION
**Pull request by Vatsal23082000**
- Added the Ephermeral disk storage size value to IMPALA_DATACACHE_CAPACITY_PARAM.

Testing
- Covered changes with unit tests.

See detailed description in the commit message.

### CODE REVIEW

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/yarn/YarnVolumeConfigProviderTest.java [hunk 1]
```diff
@@ -70,7 +70,7 @@ void getRoleConfigsWithMultipleEphemeralVolumes() {
     @Test
     void getRoleConfigsWithMultipleAttachedVolumes() {
         HostgroupView worker = hostGroupWithVolumeTemplatesAndTemporaryStorage(2,
-                Sets.newHashSet(new VolumeTemplate()), TemporaryStorage.ATTACHED_VOLUMES, 0);
+                Sets.newHashSet(new VolumeTemplate()), TemporaryStorage.ATTACHED_VOLUMES, 0, 00);
```
Comment (bergerdenes) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/yarn/YarnVolumeConfigProviderTest.java:73]: double 0

File: core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-- // CB-17015_Add_InstanceStorageSize_to_Template
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE template ADD COLUMN IF NOT EXISTS instancestoragesize INTEGER;
+UPDATE template SET instancestoragesize = 0 WHERE instancestoragesize IS NULL;
```
Comment (bergerdenes) [inline core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql:5]: Do you have stats on how many records are affected?
Please avoid setting defaults for new fields (as well as NON NULL constraints). Solve it from backend code.

File: core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+ALTER TABLE template ALTER instancestoragesize SET DEFAULT 0;
```
Comment (bergerdenes) [inline core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql:6]: please remove this

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 4]
```diff
@@ -38,20 +40,51 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
+                        temporaryStorageVolumeSize = hostGroupView.getTemporaryStorageVolumeSize();
+                    }
+                }
 
                 List<ApiClusterTemplateConfig> configs = new ArrayList<>();
+
+                if (checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
+                    configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
+                            buildEphemeralVolumePathString(temporaryStorageVolumeCount, "impala/scratch")));
+                } else {
```
Comment (bergerdenes) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:61]: wrong indent below

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 5]
```diff
@@ -38,20 +40,51 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
                         .mapToInt(volume -> volume.getVolumeSize())
                         .min()
                         .orElse(0);
+                Integer volumeCount = 0;
+                Integer temporaryStorageVolumeCount = 0;
+                Integer temporaryStorageVolumeSize = 0;
```
Comment (bergerdenes) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:45]: Can be primitive `int`s

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 6]
```diff
@@ -38,20 +40,51 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
 
                     configs.add(config(IMPALA_DATACACHE_ENABLED_PARAM, "true"));
-                    configs.add(config(IMPALA_DATACACHE_CAPACITY_PARAM, Long.toString(dataCacheCapacityInBytes)));
-                    configs.add(config(IMPALA_DATACACHE_DIRS_PARAM,
-                            buildVolumePathStringZeroVolumeHandled(hostGroupView.getVolumeCount(), "impala/datacache")));
+
+                    if (checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
+                        long temporaryStoragedataCacheCapacityInBytes = VolumeUtils
+                                .convertGBToBytes(Math.min(temporaryStorageVolumeSize / 2,
+                                        MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / hostGroupView.getTemporaryStorageVolumeCount()));
```
Comment (bergerdenes) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:76]: why not using the local variable?

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/yarn/YarnVolumeConfigProviderTest.java (hunk 1 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/yarn/YarnVolumeConfigProviderTest.java:73]: Yes, I will change it to single 0. Made a mistake with it.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java (hunk 4 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:61]: Yes, I will change the indent.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java (hunk 6 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:76]: Yes, now local variable can be used. First, I had not created the local variable. So, I was calling it directly. I will change it.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java (hunk 5 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:45]: Yes, we can use primitive int. I will change it for both of them.

File: core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql (hunk 3 above)
Comment (Vatsal23082000) [reply core/src/main/resources/schema/app/20220727094806_CB-17015_Add_InstanceStorageSize_to_Template.sql]: Yes, I have removed them in my commit.

File: cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/model/InstanceStoreMetadata.java [hunk 7]
```diff
@@ -28,6 +28,11 @@ public Integer mapInstanceTypeToInstanceStoreCountNullHandled(String instanceTyp
         return instanceStoreCount != null ? instanceStoreCount : 0;
     }
 
+    public Integer mapInstanceTypeToInstanceSizeNullHandled(String instanceType) {
+        Integer instanceSize = instaceStoreConfigMap.getOrDefault(instanceType, VolumeParameterConfig.EMPTY).minimumSize();
```
Comment (nadamdb) [inline cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/model/InstanceStoreMetadata.java:32]: Did you test the functionality on Azure as well as AWS? AFAIK there are Azure instance types where besides the one ephemeral volume that comes with every Azure instance there are also additional NVMW ephemeral volumes that can be of different size than the default one ephemeral volume. And I think Azure implementation of the instance store metadata collection is not prepared for this scenario and only returns info about the one default ephemeral storage

Comment (Vatsal23082000) [reply cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/model/InstanceStoreMetadata.java:32]: Yes, I had discussed this with @juhi-09. So, we had decided to remove this patch for AWS first and after that would look for changes in the Azure part.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 8]
```diff
@@ -38,20 +40,51 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
+                if (checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
+                    configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
+                            buildEphemeralVolumePathString(temporaryStorageVolumeCount, "impala/scratch")));
+                } else {
+                    configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
+                            buildVolumePathStringZeroVolumeHandled(volumeCount, "impala/scratch")));
+                }
 
                 if (minAttachedVolumeSize > MIN_ATTACHED_VOLUME_SIZE_TO_ENABLE_CACHE_IN_GB
                         && hostGroupView.getVolumeCount() > 0) {
```
Comment (bergerdenes) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:67]: `volumeCount` is already extracted above

**Review by bergerdenes — CHANGES_REQUESTED** [review]
You will need to rebase on top of master too.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java (hunk 8 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:67]: Yes, I will change it here also.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 9]
```diff
@@ -38,20 +40,49 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
+                    configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
+                            buildVolumePathStringZeroVolumeHandled(volumeCount, "impala/scratch")));
+                }
 
                 if (minAttachedVolumeSize > MIN_ATTACHED_VOLUME_SIZE_TO_ENABLE_CACHE_IN_GB
-                        && hostGroupView.getVolumeCount() > 0) {
+                        && volumeCount > 0) {
                     long dataCacheCapacityInBytes = VolumeUtils
-                            .convertGBToBytes(Math.min(minAttachedVolumeSize / 2, MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / hostGroupView.getVolumeCount()));
+                            .convertGBToBytes(Math.min(minAttachedVolumeSize / 2, MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / volumeCount));
```
Comment (nadamdb) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:67]: It's unnecessary to declare and calculate dataCacheCapacityInBytes here as it's not used outside of the else branch in the if-else beneath (line 80). Please move it there.

**Review by nadamdb — CHANGES_REQUESTED** [review]

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java (hunk 9 above)
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:67]: Okay, I will move it.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 10]
```diff
@@ -38,20 +40,49 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
+                        temporaryStorageVolumeCount = hostGroupView.getTemporaryStorageVolumeCount();
+                        temporaryStorageVolumeSize = hostGroupView.getTemporaryStorageVolumeSize();
+                    }
+                }
 
                 List<ApiClusterTemplateConfig> configs = new ArrayList<>();
-                configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
-                        buildVolumePathStringZeroVolumeHandled(hostGroupView.getVolumeCount(), "impala/scratch")));
+
+                if (source.getCloudPlatform().equalsIgnoreCase("AWS") && checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java]: I have changed it.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 11]
```diff
@@ -38,20 +40,49 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
                     long dataCacheCapacityInBytes = VolumeUtils
-                            .convertGBToBytes(Math.min(minAttachedVolumeSize / 2, MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / hostGroupView.getVolumeCount()));
+                            .convertGBToBytes(Math.min(minAttachedVolumeSize / 2, MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / volumeCount));
 
                     configs.add(config(IMPALA_DATACACHE_ENABLED_PARAM, "true"));
-                    configs.add(config(IMPALA_DATACACHE_CAPACITY_PARAM, Long.toString(dataCacheCapacityInBytes)));
-                    configs.add(config(IMPALA_DATACACHE_DIRS_PARAM,
-                            buildVolumePathStringZeroVolumeHandled(hostGroupView.getVolumeCount(), "impala/datacache")));
+
+                    if (source.getCloudPlatform().equalsIgnoreCase("AWS") && checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java]: I have changed it.

**Comment by Vatsal23082000** [comment]
I have made the changes for the azure part. The below images are for the Data Mart Cluster in Azure where Ephemeral volumes are being used.
resolve 1
resolve 2

 The below images are for the Data Mart Cluster in Azure where Ephemeral volumes are not being used and the attached volumes are being used.
 
resolve3
resolve4

The node types which are used in DM are:
1.  master - Standard_E8_v3
2.  coordinator/executor - Standard_E16_v3

The node types which are used in RT-DM are:
1.  master - Standard_D16_v3
2.  coordinator/executor - Standard_D16_v3

**Review by nadamdb — CHANGES_REQUESTED** [review]

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java [hunk 12]
```diff
@@ -38,20 +40,49 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
...
                         .min()
                         .orElse(0);
+                int volumeCount = 0;
+                int temporaryStorageVolumeCount = 0;
+                int temporaryStorageVolumeSize = 0;
+                if (hostGroupView != null) {
+                    volumeCount = hostGroupView.getVolumeCount();
+                    if (hostGroupView.getTemporaryStorageVolumeCount() != null) {
+                        temporaryStorageVolumeCount = hostGroupView.getTemporaryStorageVolumeCount();
+                        temporaryStorageVolumeSize = hostGroupView.getTemporaryStorageVolumeSize();
```
[the remark this answers is not in the document]
Comment (Vatsal23082000) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java:50]: Okay, will change it.

**Review by juhi-09 — APPROVED** [review]
LGTM

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**